File: 动态规划/198.打家劫舍.go
func rob(nums []int) int {
	n := len(nums)
	memo := make([]int, n)
	for i := 0; i < n; i++ {
		memo[i] = -1
	}
	return dfs(n-1, nums, memo)
}

func dfs(i int, nums []int, memo []int) int {
	if i < 0 {
		return 0
	}
	if memo[i] != -1 {
		return memo[i]
	}
	node1 := dfs(i-1, nums, memo)
	node2 := dfs(i-2, nums, memo) + nums[i]
	memo[i] = max(node1, node2)
	return memo[i]
}

File: 动态规划/198.打家劫舍.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        int[] memo = new int[n];
        Arrays.fill(memo,-1);
        return dfs(nums,n-1,memo);
    }

    int dfs(int[] nums,int i,int[] memo){
        if(i<0){
            return 0;
        }
        if(memo[i]!=-1){
            return memo[i];
        }
        return memo[i]=Math.max(dfs(nums,i-1,memo),dfs(nums,i-2,memo)+nums[i]);
    }
}